src/lib/core/datetime.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <string.h>
+#include <time.h>
+
+#include "c-dt/dt.h"
+#include "trivia/util.h"
+#include "datetime.h"
+
+/**
+ * Given the seconds from Epoch (1970-01-01) we calculate date
+ * since Rata Die (0001-01-01).
+ * DT_EPOCH_1970_OFFSET is the distance in days from Rata Die to Epoch.
+ */
+static int
+local_dt(int64_t secs)
+{
+	return dt_from_rdn((int)(secs / SECS_PER_DAY) + DT_EPOCH_1970_OFFSET);
+}
+
+static int64_t
+local_secs(const struct datetime *date)
+{
+	return (int64_t)date->epoch + date->tzoffset * 60;
+}
+
+static void
+datetime_to_tm(const struct datetime *date, struct tm *tm)
+{
+	memset(tm, 0, sizeof(*tm));
+	int64_t local_epoch = local_secs(date);
+	dt_to_struct_tm(local_dt(local_epoch), tm);
+
+	tm->tm_gmtoff = date->tzoffset * 60;
+
+	int seconds_of_day = local_epoch % SECS_PER_DAY;
+	tm->tm_hour = (seconds_of_day / 3600) % 24;
+	tm->tm_min = (seconds_of_day / 60) % 60;
+	tm->tm_sec = seconds_of_day % 60;
+}
+
+void
+tnt_datetime_now(struct datetime *now)
+{
+	struct timeval tv;
+	gettimeofday(&tv, NULL);
+	now->epoch = tv.tv_sec;
+	now->nsec = tv.tv_usec * 1000;
+
+	struct tm tm;
+	localtime_r(&tv.tv_sec, &tm);
+	now->tzoffset = tm.tm_gmtoff / 60;
+}
+
+size_t
+tnt_datetime_strftime(const struct datetime *date, char *buf, size_t len,
+		      const char *fmt)
+{
+	struct tm tm;
+	datetime_to_tm(date, &tm);
+	return strftime(buf, len, fmt, &tm);
+}
+
+/**
+ * NB! buf may be NULL, and we should handle it gracefully, returning
```

Comment:
Test of `snprintf` functionality is part of messagepack commit, please see it here, https://github.com/tarantool/tarantool/pull/6426/commits/bbf4a9e2ac97173c58d275fc819c6157ce105d8c#diff-60e9c18f243e194504bcf18af0f0d4bfa1e9be63add7e465630e85f98636a443R161-R174

I would rather not bring it here, as it's not part of RFC for stage.1

Reply:
I know, it is probably used more extensively later. But the code is here now. Lets cover it with tests now then. Otherwise you bring dead code. If we submit these first 3 patches first, they will bring untested code. And who knows when we will add the tests with the next commits.